cumulus/client/bootnodes/src/discovery.rs

```diff
@@ -0,0 +1,460 @@
+	) {
+		let direct_request = self.direct_requests.remove(&peer_id);
+
+		let response = match res {
+			Ok(Ok((payload, _))) => match Response::decode(payload.as_slice()) {
+				Ok(response) => response,
+				Err(e) => {
+					warn!(
+						target: LOG_TARGET,
+						"Failed to decode parachain bootnode response from {peer_id:?}: {e}",
+					);
+					return;
+				},
+			},
+			Ok(Err(e)) => {
+				if direct_request {
+					// It only makes sense to try to find the node on the DHT in case of "address
+					// not available" error. Unfortunately, libp2p and litep2p backends report such
+					// errors differently, and also some network library could break the error
+					// reporting in the future. So, to be on the safe side and avoid subtle bugs,
+					// we always try to find the node on the DHT in case of the request failure.
+					debug!(
+						target: LOG_TARGET,
+						"Failed to directly query parachain bootnode from {peer_id:?}: {e}. \
+						 Starting FIND_NODE query on the DHT",
+					);
+					self.find_node_queries.insert(peer_id);
+					self.relay_chain_network.find_closest_peers(peer_id);
+				} else {
+					debug!(
+						target: LOG_TARGET,
+						"Failed to query parachain bootnode from {peer_id:?} after finding
+						 the node addresses on the DHT: {e}",
+					);
+				}
+				return;
+			},
+			Err(_) => {
+				debug!(
+					target: LOG_TARGET,
+					"Parachain bootnode request to {peer_id:?} canceled. \
+					 The node is likely terminating.",
+				);
+				return;
+			},
+		};
+
+		match (response.genesis_hash, response.fork_id) {
+			(genesis_hash, fork_id)
+				if genesis_hash == self.parachain_genesis_hash &&
+					fork_id == self.parachain_fork_id => {},
+			(genesis_hash, fork_id) => {
+				warn!(
+					target: LOG_TARGET,
+					"Received invalid parachain bootnode response from {peer_id:?}: \
+					 genesis hash {}, fork ID {:?} don't match expected genesis hash {}, fork ID {:?}",
+					hex::encode(genesis_hash),
+					fork_id,
+					hex::encode(self.parachain_genesis_hash.clone()),
+					self.parachain_fork_id,
+				);
+				return;
+			},
+		}
+
+		let paranode_peer_id = match PeerId::from_bytes(response.peer_id.as_slice()) {
+			Ok(peer_id) => peer_id,
+			Err(e) => {
+				warn!(
+					target: LOG_TARGET,
+					"Failed to decode parachain peer ID in response from {peer_id:?}: {e}",
+				);
+				return;
+			},
+		};
+
+		if paranode_peer_id == self.parachain_network.local_peer_id() {
+			warn!(
+				target: LOG_TARGET,
+				"Received own parachain node peer ID in bootnode response from {peer_id:?}. \
+				 This should not happen as we don't request parachain bootnodes from self.",
+			);
+			return;
+		}
+
+		let paranode_addresses = response
+			.addrs
+			.into_iter()
+			.map(Multiaddr::try_from)
+			.collect::<Result<Vec<_>, _>>();
+		let paranode_addresses = match paranode_addresses {
+			Ok(paranode_addresses) => paranode_addresses,
+			Err(e) => {
```

Comment:
If the node provides one invalid address, we'll discard all addresses here

Could it be possible that the peer has a few wrong addresses (ie discovered wrongly by /identify or kad) increasing the likely hood of bootnodes discovery failures?

Reply:
It is unlikely other addresses were serialized properly if the one has invalid binary representation.